csrc/scheduler/reduction.cpp

```diff
@@ -521,15 +521,32 @@ std::shared_ptr<ReductionParams> outerReductionHeuristic(
   const int64_t device_multiprocessor_count =
       (int64_t)at::cuda::getCurrentDeviceProperties()->multiProcessorCount;
 
+  // Set register used to store vectorized and unrolled data loaded from gmem.
+  // A large value allows more unroll and vectorization, which is beneficial
+  // for memory-bound kernels. However, it increases register pressure and may
+  // lead to lower occupancy which is bad for compute-bound kernels. In most
+  // cases, the scheduler uses 512 threads and to reach an occupancy of 50%,
+  // each thread can use up to 64 registers, here only 8 registers are reserved
+  // for unroll and vectorization. The fused ops can have 48 registers for other
+  // purposes. Test shows it leads to 50% occupancy for outer reduction without
+  // fused ops and 50% occupancy for gelu backward which fused 21 ops including
+  // the expensive tanh op. Further tuning of this heuristic can utilize the
+  // cost of the fused ops.
+  const int64_t buffer_reg_count = 8L;
```

Comment:
nitpick, PR description mentioned about increasing max_unroll from 8 to 16. I imagine that's assuming data type fp16.

I'm curious on how this would translate to fp32 in terms of performance, are all the perf measurement done in fp32?

Reply:
heatmap is based on fp16, cpp benchmark has both fp16 & fp32.

Reply:
interesting. I'm wondering if cpp benchmark regressions are just catching the small reduction tensors or is it showing regression on fp32.

Do you think we should get the fp32 benchmark with the heatmap configs just to verify that?

Reply:
CPP benchmarks didn't show regression for fp32.
Double checked heatmap test with fp32, also no regression.
**Outer reduction**
![image](https://github.com/NVIDIA/Fuser/assets/116412316/64870008-7c24-4494-bd8b-7ec37430e2b0)
**Gelu Bwd**
![image](https://github.com/NVIDIA/Fuser/assets/116412316/ff64444a-bcf0-4527-9240-38e8fb91243a)